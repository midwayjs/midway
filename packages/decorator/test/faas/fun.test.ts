import {
  Func,
  FUNC_KEY,
  getClassMetadata,
  getObjectDefProps,
  listModule,
  ScopeEnum,
  ServerlessFunction,
  ServerlessTrigger,
  ServerlessTriggerType,
} from '../../src';

@Func('index.handler', { middleware: ['hello'] })
class TestFun {
}

class TestFun1 {
  @Func('ttt.handler')
  @ServerlessFunction({
    functionName: 'abcde'
  })
  @ServerlessTrigger(ServerlessTriggerType.HTTP, { method: 'get', path: '/' })
  @ServerlessTrigger(ServerlessTriggerType.API_GATEWAY, { method: 'get', path: '/' })
  @ServerlessTrigger(ServerlessTriggerType.MTOP)
  @ServerlessTrigger(ServerlessTriggerType.HSF)
  @ServerlessTrigger(ServerlessTriggerType.EVENT)
  @ServerlessTrigger(ServerlessTriggerType.CDN)
  @ServerlessTrigger(ServerlessTriggerType.LOG, {
    source: '',
    project: '',
    log: ''
  })
  @ServerlessTrigger(ServerlessTriggerType.MQ, {
    topic: '',
    tags: '',
    region: '',
    strategy: 'BACKOFF_RETRY'
  })
  @ServerlessTrigger(ServerlessTriggerType.OS, {
    bucket: '',
    events: '',
    filter: {
      prefix: '',
      suffix: ''
    }
  })
  @ServerlessTrigger(ServerlessTriggerType.TIMER, {
    type: 'every',
    value: '5m',
  })
  invoke() {
  }
}

describe('/test/faas/fun.test.ts', () => {
  it('fun decorator should be ok', () => {
    const meta = getClassMetadata(FUNC_KEY, TestFun);
    expect(meta).toStrictEqual([
      {
        funHandler: 'index.handler',
        middleware: ['hello'],
      },
    ]);

    const c = getClassMetadata(FUNC_KEY, TestFun1);
    expect(c).toMatchSnapshot();

    const def = getObjectDefProps(TestFun);
    expect(def).toStrictEqual({
      scope: ScopeEnum.Request,
    });

    const m = listModule(FUNC_KEY);
    expect(m.length).toEqual(2);
  });
});
