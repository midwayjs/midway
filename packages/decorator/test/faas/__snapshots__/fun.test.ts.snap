// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/faas/fun.test.ts fun decorator should be ok 1`] = `
Array [
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
      "type": "every",
      "value": "5m",
    },
    "methodName": "invoke",
    "type": "timer",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "bucket": "",
      "events": "",
      "filter": Object {
        "prefix": "",
        "suffix": "",
      },
      "functionName": "testFun1-invoke",
    },
    "methodName": "invoke",
    "type": "os",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
      "region": "",
      "strategy": "BACKOFF_RETRY",
      "tags": "",
      "topic": "",
    },
    "methodName": "invoke",
    "type": "mq",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
      "log": "",
      "project": "",
      "source": "",
    },
    "methodName": "invoke",
    "type": "log",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
    },
    "methodName": "invoke",
    "type": "cdn",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
    },
    "methodName": "invoke",
    "type": "event",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
    },
    "methodName": "invoke",
    "type": "hsf",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
    },
    "methodName": "invoke",
    "type": "mtop",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
      "method": "get",
      "path": "/",
    },
    "methodName": "invoke",
    "type": "apigw",
  },
  Object {
    "functionName": "testFun1-invoke",
    "metadata": Object {
      "functionName": "testFun1-invoke",
      "method": "get",
      "path": "/",
    },
    "methodName": "invoke",
    "type": "http",
  },
  Object {
    "descriptor": Object {
      "configurable": true,
      "enumerable": false,
      "value": [Function],
      "writable": true,
    },
    "funHandler": "ttt.handler",
    "key": "invoke",
  },
]
`;
